cmd/xfn/spark/spark.go

```diff
@@ -0,0 +1,217 @@
+limitations under the License.
+*/
+
+// Package spark runs a Composition Function. It is designed to be run as root
+// inside an unprivileged user namespace.
+package spark
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"time"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/uuid"
+	runtime "github.com/opencontainers/runtime-spec/specs-go"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+
+	"github.com/crossplane/crossplane/apis/apiextensions/fn/proto/v1alpha1"
+	"github.com/crossplane/crossplane/internal/oci"
+	"github.com/crossplane/crossplane/internal/oci/spec"
+	"github.com/crossplane/crossplane/internal/oci/store"
+	"github.com/crossplane/crossplane/internal/oci/store/overlay"
+	"github.com/crossplane/crossplane/internal/oci/store/uncompressed"
+)
+
+// Error strings.
+const (
+	errReadRequest      = "cannot read request from stdin"
+	errUnmarshalRequest = "cannot unmarshal request data from stdin"
+	errNewBundleStore   = "cannot create OCI runtime bundle store"
+	errNewDigestStore   = "cannot create OCI image digest store"
+	errParseRef         = "cannot parse OCI image reference"
+	errPull             = "cannot pull OCI image"
+	errBundleFn         = "cannot create OCI runtime bundle"
+	errMkRuntimeRootdir = "cannot make OCI runtime cache"
+	errRuntime          = "OCI runtime error"
+	errCleanupBundle    = "cannot cleanup OCI runtime bundle"
+	errMarshalResponse  = "cannot marshal response data to stdout"
+	errWriteResponse    = "cannot write response data to stdout"
+	errCPULimit         = "cannot limit container CPU"
+	errMemoryLimit      = "cannot limit container memory"
+	errHostNetwork      = "cannot configure container to run in host network namespace"
+)
+
+// The path within the cache dir that the OCI runtime should use for its
+// '--root' cache.
+const ociRuntimeRoot = "runtime"
+
+// The time after which the OCI runtime will be killed if none is specified in
+// the RunFunctionRequest.
+const defaultTimeout = 25 * time.Second
+
+// Command runs a containerized Composition Function.
+type Command struct {
+	CacheDir string `short:"c" help:"Directory used for caching function images and containers." default:"/xfn"`
+	Runtime  string `help:"OCI runtime binary to invoke." default:"crun"`
+}
+
+// Run a Composition Function inside an unprivileged user namespace. Reads a
+// protocol buffer serialized RunFunctionRequest from stdin, and writes a
+// protocol buffer serialized RunFunctionResponse to stdout.
+func (c *Command) Run() error { //nolint:gocyclo // TODO(negz): Refactor some of this out into functions, add tests.
+	pb, err := io.ReadAll(os.Stdin)
+	if err != nil {
+		return errors.Wrap(err, errReadRequest)
+	}
+
+	req := &v1alpha1.RunFunctionRequest{}
+	if err := proto.Unmarshal(pb, req); err != nil {
+		return errors.Wrap(err, errUnmarshalRequest)
+	}
+
+	t := req.GetRunFunctionConfig().GetTimeout().AsDuration()
```

Comment:
Would this crash if timeout is not set, e.g. GetTimeout returns nil?

Reply:
No, I don't think so. Most Go code generated from protobufs is pretty good about nil checking for you. `GetTimeout()` returns a `*durationpb.Duration`. You can see that `AsDuration()` calls `GetSeconds()` and `GetNano()`, both of which will just return 0 if the `*Duration` is nil.

https://github.com/protocolbuffers/protobuf-go/blob/v1.28.1/types/known/durationpb/duration.pb.go#L176